### ISSUE TITLE
Timezone support

### ISSUE THREAD

**Issue by Nevon** [calendso/calendso#17]
Currently it looks like the calendar is hard-coded to `Europe/London`. This should likely be set to the timezone of the calendar, or perhaps a setting on the user.

**Comment by abrambailey** [comment]
Yeah, this won't work for me either without it.

**Comment by emrysal** [comment]
https://github.com/calendso/calendso/pull/8

Currently this translates it without telling the user the person they are booking is in a different timezone, so this would be a future addition. Current behaviour shows the bookable slots as it is known by the user, so for example a meeting slot booked for 3pm agent time UTC+0600 would show up as a 9am slot for the booker.

**Comment by abrambailey** [comment]
Interesting, how do you know the user's TZ? What if they are traveling?

**Comment by surjithctly** [comment]
@abrambailey Usually it's based on Geo-location along with an option to modify it. 

I prefer something like this:

![image](https://user-images.githubusercontent.com/1884712/114890743-254cbd80-9dc0-11eb-923a-577245fdc23c.png)

**Comment by Nevon** [comment]
You can get the offset between the current user and UTC by doing `new Date().getTimezoneOffset()`. I'm in CET, which is currently two hours ahead of UTC, which means that it returns `-120`.

**Comment by abrambailey** [comment]
I would place the time zone drop down here, as people may miss it on the first page:

![image](https://user-images.githubusercontent.com/978462/114897700-19b8c100-9dd7-11eb-8726-24e79b535846.png)

**Comment by 50bbx** [comment]
Hey guys,

I implemented this feature, both for the owner of the calendar and for the user choosing a slot.

No dropdown yet (just getting default timezone), but I will create a PR tomorrow.

**Comment by 50bbx** [comment]
Basic time zone support PR: https://github.com/calendso/calendso/pull/28

**Comment by baileypumfleet** [comment]
Okay, so we have the basic timezone support PR merged in and everything, so let's discuss next steps. I know we would still need to add the timezone label or dropdown on the booking page, for instance.

**Comment by surjithctly** [comment]
when will this available? I don't see it in the live demo.

**Comment by baileypumfleet** [comment]
Just to update you all, as part of this, I've added a new field on the booking screen's left column which shows the guessed timezone. If you click it, there's a dropdown to switch between 12 and 24 hours, but that's where the timezone selection should also go.

**Comment by surjithctly** [comment]
Perfect.

![image](https://user-images.githubusercontent.com/1884712/115426561-7809e900-a1b5-11eb-8f6e-71d7ba415fd7.png)

**Comment by 50bbx** [comment]
@baileypumfleet Did you make a pull request?

I have the code ready locally to switch between timezones

**Comment by baileypumfleet** [comment]
@50bbx No, I directly committed it

**Comment by soneji** [comment]
Hi all, I just thought I'd drop by and mention a test case that often causes a problem with timezones, and has also caused a problem for [Easy Appointments](https://easyappointments.org/)

The test case is for countries/timezones that change for daylight saving time. It can often result in entire calendars being out of sync by 1 hour and also possibly one week with correct timings, and the week after daylight saving time starts being 1 hour out of sync

Thanks in advance

**Comment by baileypumfleet** [comment]
@50bbx Is your code for switching between the timezones ready for a PR? Super excited to get this added in :)

**Comment by 50bbx** [comment]
I saw that there were some changes on the timezone component selector, [here](https://github.com/calendso/calendso/issues/80) We need to address those and understand how we are going to manage the timezone selector.

I need to add that component to the book an appointment page.

When we settle things down on that side I will extract the timezone selector in a separate file and make the PR.

**Comment by harad1** [comment]
> Hi all, I just thought I'd drop by and mention a test case that often causes a problem with timezones, and has also caused a problem for [Easy Appointments](https://easyappointments.org/)
> 
> The test case is for countries/timezones that change for daylight saving time. It can often result in entire calendars being out of sync by 1 hour and also possibly one week with correct timings, and the week after daylight saving time starts being 1 hour out of sync
> 

We could go think about using [Coordinated Universal Time](https://support.google.com/calendar/answer/37064?co=GENIE.Platform%3DAndroid&hl=en), like Google Calendar does, to not run into the issue of daylight saving. 

[https://en.wikipedia.org/wiki/Coordinated_Universal_Time](https://en.wikipedia.org/wiki/Coordinated_Universal_Time)

**Comment by emrysal** [comment]
Hey @50bbx - This was a small amend not to change any logic, but whilst condensing the options I was thinking that it would make sense to extract the timezone logic into a seperate component as well, to implement extra features. For example so we can also show the GMT+X difference as part of the label, then include this component similarly on the appointment section. 

@harad1 Internally, @50bbx and I used UTC; you can see this for example by the call to /book (which is all done in UTC). UTC works with all calendars regardless of the timezone. Even though this works theoretically; it needs double checking. In addition, the reason there is some options in the dropdown selector that seem awfully specific (Like "Europe/Bucharest") this has the reason that they handle DST differently; Bucharest for example enters and leaves DST one hour later (at 3am) than the rest of Europe (mostly 2am).

**Comment by 50bbx** [comment]
Ok, I tried to add the TimeZone selector used [here](https://github.com/calendso/calendso/issues/103). But now it is really hard to use it with `dayjs.tz.guess()`.

In the `pages/[user]/[type].tsx` file these are the critical changes:

```js
export default function Type(props) {
  //...
  const [selectedTimeZone, setSelectedTimeZone] = useState(dayjs.tz.guess());
  //...
  return (
    <TimezoneSelect id="timeZone" value={selectedTimeZone} onChange={setSelectedTimeZone} className="shadow-sm focus:ring-blue-500 focus:border-blue-500 mt-1 block w-full sm:text-sm border-gray-300 rounded-md" />
  )
}
```
In my case `dayjs.tz.guess()` returns `Europe/Rome`, but `react-timezone-select` aggregates timezones and `Europe/Rome` is not present, only `Europe/Amsterdam` and `Europe/Brussels` are.

Basically, setting value this way, defaulting to the `dayjs.tz.guess()` for the selector does not work. This means that if we use this component on the appointment booking page, we need to find a way to aggregate timezones the same way.

**Comment by emrysal** [comment]
In reaction to that @50bbx; I opened an issue on https://github.com/ndom91/react-timezone-select/issues/25 - let's see what comes back.

**Comment by ulentini** [comment]
From what I can see, react-timezone-select support all official Timezone names (because the `spacetime` package supports them), but it provides just a subset of them (removing duplicates). 
If a complete list of timezones is passed in the `timezones` attribute, `dayjs.tz.guess()` can still be used.

**Comment by emrysal** [comment]
@50bbx With v1.0.2 this should now be possible 👍

**Comment by 50bbx** [comment]
Thanks @emrysal it works!

See https://github.com/calendso/calendso/pull/163

**Comment by abrambailey** [comment]
Just a small issue here. The first image comes from my email ... The second comes from the icon to add to google calendar in the confirmation page. Not sure which one is wrong, but they can't both be right:

![image](https://user-images.githubusercontent.com/978462/117710977-54283a80-b198-11eb-9098-7471f991e602.png)

![image](https://user-images.githubusercontent.com/978462/117711017-5d190c00-b198-11eb-9e9d-6af18bcbbca7.png)

**Comment by abrambailey** [comment]
Happy to help, has code been deployed to https://calendso.com/rick/quick-chat ?

**Comment by baileypumfleet** [comment]
Yes, should be running the latest code now.

On May 10, 2021, thekeviv ***@***.***> wrote:
> Happy to help, has code been deployed to
> <https://calendso.com/rick/quick-chat> ?
>

—
You are receiving this because you modified the open/close state.
Reply to this email directly, view it on GitHub
<https://github.com/calendso/calendso/issues/17#issuecomment-837252909>,
or unsubscribe <https://github.com/notifications/unsubscribe-
auth/AGFU7V2XPS2FCVXADBPNWITTNA4A3ANCNFSM427QJFXA>.

**Comment by jimafisk** [comment]
Can confirm I'm seeing the same behavior as @abrambailey. Behind the scenes the meeting gets added to the correct timeslot, but if you click "G" icon in the "Add to your calendar" section of the booking confirmation page, it adds it in the wrong timezone. 

<details>
<summary>Screenshot of confirmation page</summary>

![confirm](https://user-images.githubusercontent.com/5913244/119048673-8b170100-b98d-11eb-980b-f79eb152df9a.png)

</details>
<details>
<summary>Screenshot of the calendar event with incorrect time</summary>

![wrong_time](https://user-images.githubusercontent.com/5913244/119048692-8eaa8800-b98d-11eb-9e1d-9be51514c4cc.png)

</details>

I'm curious about the intention of this "add to calendar" feature if the event is already getting added to the calendar of the person based on their email address. Is it so the person requesting the appointment can add it to another calendar?

**Comment by TonBil** [comment]
Newbie (to cal.com) here. I set my VPN to a different timezone, yet the calendar shows booking times in my actual timezone. I see that the booker can actually change to their timezone in the left column. But who will see and do that? It's not intuitive. 
So, is automatic geolocation for timezone setting not yet implied as standard? That would be a bummer and a reason to abandon cal.com and go back to google calendar again for bookings. 
Or is there a way to get geolocation for timezone preselection set up?

**Comment by xcellentavi** [comment]
I am experiencing something similar to this.  My timezone is set correctly in the settings and the booking emails and calendar invites are all showing the correct timezone.  It's when I go to the booking page in the admin panel and looking at upcoming bookings, they are set to Europe/London timezone.  I tried changing settings to lock it to my timezone, but it is still showing internally as the wrong time.  Any idea if this is a bug or if I am missing something to change on my end?

Thank you!
